### DISCUSSION TITLE
Lost battery input/output energy today entities

### DISCUSSION THREAD

**Discussion by mickemartinsson** [wills106/homeassistant-solax-modbus#45] [General]
Hi,
I have lost the entities for battery input and output energy - I can no longer use Energy in Lovelace (missing flows in and out of battery). Tried to reload integration but no luck. Is this an error in my system or is there something else? It has worked perfectly before.

**Comment by infradom** [comment]
Had a quick look, on first sight these parameters seem to have disappeared from the source code, except for X1 installations. I might be wrong ...  Should be easy to fix, unless there is a reason for removing them I am not aware of

**Reply by wills106** [reply]
I was meant to remove them from Gen2 Inverters and only be there for Gen3 and above.
But I only placed them in X1 by mistake.

**Reply by wills106** [reply]
Just trying to work out how to copy "battery_input_energy_today" & "battery_output_energy_today" from GEN3_X1_SENSOR_TYPES to GEN3_X3_SENSOR_TYPES

Easiest way would to just copy the whole block over

```
	"battery_input_energy_today": SolaXModbusSensorEntityDescription(
		name="Battery Input Energy Today",
		key="input_energy_charge_today",
		native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
        icon="mdi:solar-power",
        device_class=DEVICE_CLASS_ENERGY,
        state_class=STATE_CLASS_TOTAL_INCREASING,
    ),
    "battery_output_energy_today": SolaXModbusSensorEntityDescription(
		name="Battery Output Energy Today",
		key="output_energy_charge_today",
		native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
        icon="mdi:solar-power",
        device_class=DEVICE_CLASS_ENERGY,
        state_class=STATE_CLASS_TOTAL_INCREASING,
    ),
```

But there must be an easier way similar to `GEN4_X1_SENSOR_TYPES = GEN3_X1_SENSOR_TYPES.copy()`
But only copying the two dictionary items needed "battery_input_energy_today" & "battery_output_energy_today"

I have tried things like `GEN3_X3_SENSOR_TYPES = GEN3_X1_SENSOR_TYPES.get("battery_input_energy_today")` but not 100% sure what I am doing.

**Reply by infradom** [reply]
Yes, I already made these changes in my pull request (not in an intelligent way)
I also added the same total variables as you will see. These totals make sense to estimate the return on investment of the batteries

**Reply by infradom** [reply]
maybe we should put them all in one of the dictionaries, then do the copy() and then remove the unneeded with pop() statements

**Reply by wills106** [reply]
Isn't `["input_energy_charge"] = round((input_energy_charge_msb * 256*256 + input_energy_charge_lsb) * 0.1, 1)` jus the same as `input_energy_charge_today` on the Gen3 and Gen4?

If it is the same I probably isn't much use for the Gen3 and Gen4 although it might be handy on the Gen2 as it doesn't have `input_energy_charge_today` although the msb and lsb are in a slight different address on the Gen2

**Reply by wills106** [reply]
> maybe we should put them all in one of the dictionaries, then do the copy() and then remove the unneeded with pop() statements

Might be worthwhile as it's all becoming a bit of a mess to be honest.

**Reply by infradom** [reply]
Yes, it is worthwile, but may take some time if we want to avoid errors
Shall I try to do this cleanup/restructuring or do you prefer to do it yourself? I may not be able to find time before wednesday

**Reply by wills106** [reply]
I don't mind. There is no rush for it though.

Probably need to make a sensor list for each Inverter. Then we can make sure we don't loose a sensor or add in wrong sensors.

**Reply by infradom** [reply]
or create list of all sensors, but extended with flags that indicate which generations (and type: X1 or X3) it applies to.
Then we can create the entities based on the inverter type and generation in a simple python loop

**Reply by wills106** [reply]
So something a bit like:

"pv_total_power"gen2,gen3,gen4
"discharger_end_time_1"gen2,gen4
"feedine_nightcharge_upper_soc"gen4

**Comment by infradom** [comment]
no, it is not the same, today gives the numbeers for the last day, these new variables give the totals since day zero, so it gives me an indication how much money my batteries have saved since day 1

**Reply by wills106** [reply]
Ahh ok

**Reply by wills106** [reply]
So it might still give battery dashboard support for the Gen2 then. Increasing total. Just need to adapt the registers for the Gen2

**Comment by infradom** [comment]
As an example: In my system, the 'today' value returns around 7kWh, the
total value returns around 800 kWh

On Sat, Apr 2, 2022 at 1:08 PM wills106 ***@***.***> wrote:

> So it might still give battery dashboard support for the Gen2 then.
> Increasing total. Just need to adapt the registers for the Gen2
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wills106/homsassistant-solax-modbus/discussions/45#discussioncomment-2490816>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2B23WHZRJP5TW7345AO43VDATDRANCNFSM5SKTXZFA>
> .
> You are receiving this because you commented.Message ID:
> <wills106/homsassistant-solax-modbus/repo-discussions/45/comments/2490816@
> github.com>
>


-- 
Jan Colpaert
+32 478 882435

**Reply by wills106** [reply]
Hopefully we can add battery for the Gen2 into the Energy dashboard.
OutputEnergy_Charge1.LSB & OutputEnergy_Charge1.MSB are the same registers between the Gen2 and Later Models.
But the InputEnergy_Charge1.LSB & InputEnergy_Charge1.MSB equivalents have moved by 1 register.

It will mess the Energy dashboard up on the first day for a Gen2 user as all of a sudden they will have a massive value in the hundreds of kWh. But going forward they should get a similar result to our input_energy_charge_today & output_energy_charge_today

Might also be handy to expose BatterySOH for the Gen2 as I am assuming that's Battery State of Health. Although I am not sure how it calculates that %

```
Gen2
0x001D OutputEnergy_Charge1.LSB 
0x001E OutputEnergy_Charge1.MSB 
0x001F BMS warnings
0x0020 InputEnergy_Charge1.LSB
0x0021 InputEnergy_Charge1.MSB
0x0022 BatteryPackageNum
0x0023 BatterySOH

Gen3 & Gen4
0x001D OutputEnergy_Charge.LSB
0x001E OutputEnergy_Charge.MSB
0x001F BMS Warning LSB
0x0020 OutputEnergy_Charge_today
0x0021 InputEnergy_Charge.LSB
0x0022 InputEnergy_Charge.MSB
0x0023 InputEnergy_Charge_today
```

**Reply by wills106** [reply]
All being well [0.4.12b](https://github.com/wills106/homsassistant-solax-modbus/releases/tag/0.4.12b) should add battery input / output to the energy dashboard for Gen2

**Comment by infradom** [comment]
I would do something like: imaginary examlple  (still some tuning needed)
GEN2 = 0x0001
GEN3 = 0x0002
GEN4 = 0x0004
X1   = 0x0100
X3   = 0x0200
EPS  = 0x1000
ALL = 0xFFFF

ALL_ENTITIES = {

  "grid_export_total": MyEntityDescription(
     name="Grid Export Total",
     key="grid_export_total",
     native_unit_of_measurement=ENERGY_KILO_WATT_HOUR,
     icon="mdi:solar-power",
     device_class=DEVICE_CLASS_ENERGY,
     state_class=STATE_CLASS_TOTAL_INCREASING,
     entity_registry_enabled_default=False,
     ormask = GEN2 | GEN3 | GEN4
     andmask =   X1 | X3
    ),
  "backup_nightcharge_upper_soc"] = MyEntityDescription(
     name="Backup Night Charge Upper SOC",
     key="backup_nightcharge_upper_soc",
     native_unit_of_measurement=PERCENTAGE,
     device_class=DEVICE_CLASS_BATTERY,
     ormask = GEN4
     andmask = X1
    ),
 "eps_min_esc_soc": MyEntityDescription(
     name="EPS Min Esc SOC",
     key="eps_min_esc_soc",
     native_unit_of_measurement=PERCENTAGE,
     ormask = GEN2 | GEN3 | GEN4
     andmask = EPS
    ),

On Sat, Apr 2, 2022 at 1:48 PM wills106 ***@***.***> wrote:

> So something a bit like:
>
> "pv_total_power"gen2,gen3,gen4
> "discharger_end_time_1"gen2,gen4
> "feedine_nightcharge_upper_soc"gen4
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/wills106/homsassistant-solax-modbus/discussions/45#discussioncomment-2490903>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AC2B23UHQP7QQZYVNNI4FXLVDAXYTANCNFSM5SKTXZFA>
> .
> You are receiving this because you commented.Message ID:
> <wills106/homsassistant-solax-modbus/repo-discussions/45/comments/2490903@
> github.com>
>


-- 
Jan Colpaert
+32 478 882435

**Comment by mickemartinsson** [comment]
Thanks!! After the update, it works again!
